File: axon-sdk-python/axon/memory.py
from axon._base import _BaseClient, _BaseSyncClient
from axon.types import MemoryResult, MemorySearchResponse, StoredMemory


class MemoryClient(_BaseClient):

    async def store(
        self,
        content: str,
        tags: dict = None,
        scope: str = "project",
        ttl: int = None,
    ) -> StoredMemory:
        """
        Store a memory in the Axon server (async).
        """
        body = {"content": content, "scope": scope}
        if tags is not None:
            body["tags"] = tags
        if ttl is not None:
            body["ttl"] = ttl

        data = await self._request("POST", "/v1/memory/store", json=body)
        return StoredMemory(id=data["id"], created_at=data["created_at"])

    async def search(
        self,
        query: str,
        limit: int = 10,
        scope: str = None,
        min_similarity: float = 0.5,
    ) -> MemorySearchResponse:
        """
        Semantically search stored memories (async).
        """
        body = {
            "query": query,
            "limit": limit,
            "min_similarity": min_similarity,
        }
        if scope is not None:
            body["scope"] = scope

        data = await self._request("POST", "/v1/memory/search", json=body)

        results = [
            MemoryResult(
                id=r["id"],
                content=r["content"],
                tags=r.get("tags", {}),
                scope=r["scope"],
                agent_id=r["agent_id"],
                similarity=r["similarity"],
                created_at=r["created_at"],
            )
            for r in data["results"]
        ]

        return MemorySearchResponse(
            results=results,
            query=data["query"],
            total_found=data["total_found"],
        )

    async def get(self, memory_id: str) -> dict:
        """
        Get a specific memory by its ID (async).
        """
        return await self._request("GET", f"/v1/memory/{memory_id}")

    async def forget(self, memory_id: str) -> bool:
        """
        Delete a memory permanently (async).
        """
        data = await self._request("DELETE", f"/v1/memory/{memory_id}")
        return data.get("deleted", False)


class SyncMemoryClient(_BaseSyncClient):

    def store(
        self,
        content: str,
        tags: dict = None,
        scope: str = "project",
        ttl: int = None,
    ) -> StoredMemory:
        """
        Store a memory in the Axon server (sync).
        """
        body = {"content": content, "scope": scope}
        if tags is not None:
            body["tags"] = tags
        if ttl is not None:
            body["ttl"] = ttl

        data = self._request("POST", "/v1/memory/store", json=body)
        return StoredMemory(id=data["id"], created_at=data["created_at"])

    def search(
        self,
        query: str,
        limit: int = 10,
        scope: str = None,
        min_similarity: float = 0.5,
    ) -> MemorySearchResponse:
        """
        Semantically search stored memories (sync).
        """
        body = {
            "query": query,
            "limit": limit,
            "min_similarity": min_similarity,
        }
        if scope is not None:
            body["scope"] = scope

        data = self._request("POST", "/v1/memory/search", json=body)

        results = [
            MemoryResult(
                id=r["id"],
                content=r["content"],
                tags=r.get("tags", {}),
                scope=r["scope"],
                agent_id=r["agent_id"],
                similarity=r["similarity"],
                created_at=r["created_at"],
            )
            for r in data["results"]
        ]

        return MemorySearchResponse(
            results=results,
            query=data["query"],
            total_found=data["total_found"],
        )

    def get(self, memory_id: str) -> dict:
        """
        Get a specific memory by its ID (sync).
        """
        return self._request("GET", f"/v1/memory/{memory_id}")

    def forget(self, memory_id: str) -> bool:
        """
        Delete a memory permanently (sync).
        """
        data = self._request("DELETE", f"/v1/memory/{memory_id}")
        return data.get("deleted", False)

File: axon-sdk-js/src/memory.ts
// ─── Axon Protocol SDK — Memory Module ───

import type { HttpClient } from './_base.js';
import type {
  StoreMemoryRequest,
  StoreMemoryResponse,
  SearchMemoryRequest,
  SearchMemoryResponse,
  MemoryEntry,
  MemoryListResponse,
} from './types.js';

/**
 * Shared vector memory — store, search, retrieve, and delete.
 *
 * ```ts
 * await axon.memory.store({ content: 'User prefers dark mode' });
 * const results = await axon.memory.search({ query: 'user preferences' });
 * ```
 */
export class MemoryModule {
  constructor(private readonly http: HttpClient) {}

  /** Store a new memory entry with optional tags, scope, and TTL. */
  async store(request: StoreMemoryRequest): Promise<StoreMemoryResponse> {
    return this.http.post<StoreMemoryResponse>('/v1/memory/store', {
      content: request.content,
      tags: request.tags ?? {},
      scope: request.scope ?? 'project',
      ttl: request.ttl ?? null,
    });
  }

  /** Semantic search over stored memories. */
  async search(request: SearchMemoryRequest): Promise<SearchMemoryResponse> {
    return this.http.post<SearchMemoryResponse>('/v1/memory/search', {
      query: request.query,
      limit: request.limit ?? 20,
      min_similarity: request.min_similarity ?? 0.3,
    });
  }

  /** Get a specific memory entry by ID. */
  async get(memoryId: string): Promise<MemoryEntry> {
    return this.http.get<MemoryEntry>(`/v1/memory/${memoryId}`);
  }

  /** Delete a specific memory entry by ID. */
  async delete(memoryId: string): Promise<{ message: string }> {
    return this.http.delete<{ message: string }>(`/v1/memory/${memoryId}`);
  }

  /** List all memories in the current project. */
  async list(limit = 50): Promise<MemoryListResponse> {
    return this.http.get<MemoryListResponse>('/v1/memory/list', {
      limit: String(limit),
    });
  }
}
